consensus/event_loop.go

```diff
@@ -0,0 +1,908 @@
+	if fc.currentView != timedOutView {
+		return
+	}
+
+	// Signal the server that we are ready to time out
+	fc.Events <- &FastHotStuffEvent{
+		EventType:      FastHotStuffEventTypeTimeout, // The timeout event type
+		View:           timedOutView,                 // The view we timed out
+		TipBlockHash:   fc.tip.block.GetBlockHash(),  // The last block we saw
+		TipBlockHeight: fc.tip.block.GetHeight(),     // The last block we saw
+		QC:             fc.tip.block.GetQC(),         // The highest QC we have
+	}
+
+	// Cancel the timeout task. The server will reschedule it when it advances the view.
+	fc.nextTimeoutTask.Cancel()
+}
+
+// Evict all locally stored votes and timeout messages with stale views. We can safely use the current
+// view to determine what is stale. The consensus mechanism will never construct a block with a view
+// that's lower than its current view. Consider the following:
+// - In the event the event update the chain tip, we will vote for that block and the view it was proposed in
+// - In the event we locally time out a view, we will send a timeout message for that view
+//
+// In both cases, we will never roll back the chain tip, or decrement the current view to construct a
+// conflicting block at that lower view that we have previously voted or timed out on. So we are safe to evict
+// locally stored votes and timeout messages with stale views because we expect to never use them for
+// block construction.
+//
+// The eviction works as follows:
+// - Votes: if the next block were to be a regular block with a QC aggregated from votes, then the it must
+// satisfy nextBlock.GetView() = tip.block.GetView() + 1, which means that currentView = tip.block.GetView() + 1.
+// We can safely evict all votes where vote.GetView() < currentView - 1.
+// - Timeouts: if the next block were be an empty block with a timeout QC aggregated from timeout messages,
+// then it must satisfy nextBlock.GetView() = timeout.GetView() + 1. We can safely evict all timeout messages with
+// currentView > timeout.GetView() + 1.
+func (fc *fastHotStuffEventLoop) evictStaleVotesAndTimeouts() {
+	// Evict stale vote messages
+	for blockHash, voters := range fc.votesSeenByBlockHash {
+		for _, vote := range voters {
+			if isStaleView(fc.currentView, vote.GetView()) {
+				// Each block is proposed at a known view, and has an immutable block hash. Votes are signed on the
+				// tuple (blockhash, view). So, if any vote message for the blockhash has a view that satisfies this
+				// condition, then it's guaranteed that all votes for the same block hash have satisfy this condition.
+				// We can safely evict all votes for this block hash.
+				delete(fc.votesSeenByBlockHash, blockHash)
+				break
+			}
+		}
+	}
+
+	// Evict stale timeout messages
+	for view := range fc.timeoutsSeenByView {
+		if isStaleView(fc.currentView, view) {
+			delete(fc.timeoutsSeenByView, view)
+		}
+	}
```

Comment:
This function looks right, but my logic for it was a bit different than the comment. It goes as follows:
* fc.currentView is the view that the next block is going to be proposed on
* That block's QC or AggQC consists of votes/timeouts on the previous block, which has view fc.currentView-1
* That means you can dispose any votes or timeouts with vote.GetView() <> fc.currentView-1

Is the above logic oversimplifying it? If not, I would update the comment.

Reply:
Correct. That's what the code already does. Here's the description from the comment block above:
```
// The eviction works as follows:
// - Votes: if the next block were to be a regular block with a QC aggregated from votes, then the it must
// satisfy nextBlock.GetView() = tip.block.GetView() + 1, which means that currentView = tip.block.GetView() + 1.
// We can safely evict all votes where vote.GetView() < currentView - 1.
// - Timeouts: if the next block were be an empty block with a timeout QC aggregated from timeout messages,
// then it must satisfy nextBlock.GetView() = timeout.GetView() + 1. We can safely evict all timeout messages with
// timeout.GetView() < currentView - 1.
```

Reply:
This is a simpler explanation than what's in the current comment block. Seems fine to update the comment block to this